### DISCUSSION TITLE
SQLAlchemy picks closed or wrong connection from the pool, what else I can do to debug it?

### DISCUSSION THREAD

**Discussion by vo-va** [sqlalchemy/sqlalchemy#9887] [Usage Questions]
The Flask app uses evenlet + monkey patching (maybe this is an issue) and more than 1 DB binding. Also it uses flask-sqlalchemy package.

During processing a view request code tried to get data from a different DB, in the next order mysql -> mssql -> mysql (the same instance) and in DB logs you can see that sqlalchemy picks a connection that by some reason is down, although in the first time it was working

An option pool_pre_ping, not helping

SQLAlchemy==1.4.48

I know it is old, but the app is old too, and it would be great to fix it without rewriting if this issue depends on the version sqlalchemy

```
16:55:30,246 - sqlalchemy.pool.impl.QueuePool - DEBUG- Connection <mysql.connector.connection.MySQLConnection object at 0x7f803a5cc490> checked out from pool
16:55:30,246 - sqlalchemy.pool.impl.QueuePool - DEBUG- Pool pre-ping on connection <mysql.connector.connection.MySQLConnection object at 0x7f803a5cc490>
16:55:30,248 - sqlalchemy.engine.Engine - INFO- BEGIN (implicit)
16:55:30,250 - sqlalchemy.engine.Engine - INFO- SELECT <>
16:55:30,251 - sqlalchemy.engine.Engine - INFO- [cached since 46.62s ago] {<>}
16:55:30,253 - sqlalchemy.engine.Engine - DEBUG- Col (<cols>)
16:55:30,254 - sqlalchemy.engine.Engine - DEBUG- Row (<data>)
16:55:30,257 - sqlalchemy.engine.Engine - INFO- SELECT <>
16:55:30,257 - sqlalchemy.engine.Engine - INFO- [cached since 46.55s ago] {<>}
16:55:30,259 - sqlalchemy.engine.Engine - DEBUG- Col (<cols>)
16:55:30,259 - sqlalchemy.engine.Engine - DEBUG- Row (<data>)
16:55:31,057 - sqlalchemy.pool.impl.QueuePool - DEBUG- Created new connection <pymssql.Connection object at 0x7f803a16bf00>
16:55:31,058 - sqlalchemy.engine.Engine - INFO- select @@version
16:55:31,058 - sqlalchemy.engine.Engine - INFO- [raw sql] {}
16:55:31,145 - sqlalchemy.engine.Engine - DEBUG- Col ('',)
16:55:31,145 - sqlalchemy.engine.Engine - DEBUG- Row ('Microsoft SQL Server <>\n',)
16:55:31,147 - sqlalchemy.engine.Engine - INFO- SELECT schema_name()
16:55:31,147 - sqlalchemy.engine.Engine - INFO- [generated in 0.00026s] {}
16:55:31,330 - sqlalchemy.engine.Engine - DEBUG- Col ('',)
16:55:31,330 - sqlalchemy.engine.Engine - DEBUG- Row ('dbo',)
16:55:31,687 - sqlalchemy.engine.Engine - INFO- SELECT CAST('<>)
16:55:31,687 - sqlalchemy.engine.Engine - INFO- [generated in 0.00039s] {}
16:55:31,859 - sqlalchemy.engine.Engine - DEBUG- Col ('',)
16:55:31,861 - sqlalchemy.engine.Engine - DEBUG- Row ('<>',)
16:55:32,126 - sqlalchemy.pool.impl.QueuePool - DEBUG- Connection <pymssql.Connection object at 0x7f803a16bf00> checked out from pool
16:55:32,126 - sqlalchemy.pool.impl.QueuePool - DEBUG- Connection <pymssql.Connection object at 0x7f803a16bf00> is fresh, skipping pre-ping
16:55:32,126 - sqlalchemy.engine.Engine - INFO- BEGIN (implicit)
16:55:32,129 - sqlalchemy.engine.Engine - INFO- SELECT <>
16:55:32,129 - sqlalchemy.engine.Engine - INFO- [generated in 0.00036s] {<>}
16:55:32,225 - sqlalchemy.engine.Engine - DEBUG- Col (<cols>)
16:55:32,225 - sqlalchemy.engine.Engine - DEBUG- Row (<data>)


# next lines debug from python code 
[16:55:32,227] INFO in report_setting: True                                                               # app.logger.info(db.session.is_active)
[16:55:32,227] INFO in report_setting: <sqlalchemy.engine.base.Connection object at 0x7f803a8d5970>       # app.logger.info(db.session.connection())
[16:55:32,227] INFO in report_setting: False                                                              # app.logger.info(the_conn.closed)
[16:55:32,227] INFO in report_setting: True                                                               # app.logger.info(the_conn.in_transaction())
[16:55:32,228] INFO in report_setting: False                                                              # app.logger.info(the_conn.invalidated)


16:55:32,231 - sqlalchemy.pool.impl.QueuePool - INFO- Invalidate connection <mysql.connector.connection.MySQLConnection object at 0x7f803a5cc490> (reason: OperationalError:MySQL Connection not available.)
16:55:32,231 - sqlalchemy.pool.impl.QueuePool - DEBUG- Hard-closing connection <mysql.connector.connection.MySQLConnection object at 0x7f803a5cc490>



```

**Comment by CaselIT** [comment]
Hi,

I'm not sure what's happening here, it seems the server closed the connection and the queue pool recycled it. I don't see a problem for sqlalchemy to fix.

Can you provide more context about what were you expecting sqlalchemy to do here?

> SQLAlchemy==1.4.48
> I know it is old, but the app is old too,

version 1.4 is still supported btw

**Reply by vo-va** [reply]
Hi,

The context here is next, I am trying to restore the functionality of the old project. DB is hosted in AWS RDS, so I am assuming that the service is pretty stable and this means we can exclude network connectivity issues. 

In the logs above you can see that code trying to reuse the same connection with 2 seconds difference. First, select working, but the 2nd attempt fails. I've tried to change the parameters of pool_recycle. The original value was 280 and I've tried with 58 seconds. In both cases, the issue is the same - 2nd select failing.  Also, I am testing the service and seeing this issue not after a long time, but just after the server started or restarted  i.e. I assume the connection is pretty much fresh.

On the server side, I have the next configuration for timeouts, it might be not optimal but because of the big wait and interactive timeout values, I assume that the server should not try to close connections too early. 

```
MySQL [(none)]> show global variables like "%timeout";
+----------------------------------+----------+
| Variable_name                    | Value    |
+----------------------------------+----------+
| aurora_fwd_master_idle_timeout   | 60       |
| aurora_globaldb_rpo_wait_timeout | 60       |
| connect_timeout                  | 10       |
| delayed_insert_timeout           | 300      |
| have_statement_timeout           | YES      |
| innodb_flush_log_at_timeout      | 1        |
| innodb_lock_wait_timeout         | 50       |
| innodb_rollback_on_timeout       | OFF      |
| interactive_timeout              | 28800    |
| lock_wait_timeout                | 31536000 |
| net_read_timeout                 | 30       |
| net_write_timeout                | 60       |
| rpl_stop_slave_timeout           | 31536000 |
| slave_net_timeout                | 60       |
| wait_timeout                     | 28800    |
+----------------------------------+----------+
```

The error logs on AWS RDS are empty.

Mysql driver/package is mysql-connector-python 8.0.33

If I will try to connect to MySQL in interactive mode python interpretator or Flask shell it is working.

And in the end, I am not sure what else I can do to find the reason for this issue. 

My wild guess is that maybe because the code trying to access 2 different DB in one Flask request session, some code closes the connection/session to MySQL when switching to interaction with MSSQL, but forgets to update the state of the connection in the pool. 
Or maybe eventlet monkey patching affect code somehow so the connection state information is out of sync.

> what were you expecting sqlalchemy to do here?

Well, I hoped to get some ideas on what else I can do to debug this issue and find its roots.

Maybe the quick fix/hack can be is forcing sqlalchemy to do pre-ping. In logs you can see that code decided that pre-ping is not necessary because it is assumed that the connection is fresh.

**Reply by CaselIT** [reply]
The connection is fresh line means that the pool just created that connection, so there is no need of running pre-ping.
I'm really not sure what would cause the connection to go away soon after the connection.

Regarding ideas on how to debug, try removing external libraries until you are left with a minimum set that reproduces the error. For example try removing eventlet and or flask.

**Comment by zzzeek** [comment]
the log lines here show two different pools, and we would assume the application has two separate `Engine` objects.  it's not possible for a single connection pool to have connections from multiple drivers unless the pool were very highly modified in some way, or it was given a creator function that randomly generates connections from different drivers, which is not something that should be done.

**Reply by vo-va** [reply]
My app uses flask-sqlalchemy package to bind object models to different DB. I don't expect it will do something weird with pools. 

Maybe these details are also necessary. 

app.db.session - variable is a default binding to MySQL 
And connections to MSSQL are done through binding in the different Model objects  __bind_key__ https://flask-sqlalchemy.palletsprojects.com/en/2.x/api/#models (I know it is not your project, just for reference in case it is important)
